pkg/controller/job: re-honor exponential backoff delay

nikhita opened the pull request:
#### What type of PR is this?

/kind regression

#### What this PR does / why we need it:

This commit makes the job controller honor exponential backoff for failed pods. Before this commit, the controller created pods without any backoff. This is a regression because the controller used to create pods with an exponential backoff delay before (10s, 20s, 40s ...).

The issue occurs only when the JobTrackingWithFinalizers feature is enabled (which is enabled by default right now). With this feature, we get an extra pod update event when the finalizer of a failed pod is removed.

Note that the pod failure detection and new pod creation happen in the same reconcile loop so the 2nd pod is created immediately after the 1st pod fails. The backoff is only applied on 2nd pod failure, which means that the 3rd pod is created 10s after the 2nd pod, 4th pod is created 20s after the 3rd pod and so on.

This commit fixes a few bugs:

1. Right now, each time `uncounted != nil` and the job does not see a _new_ failure, `forget` is set to true and the job is removed from the queue. Which means that this condition is also triggered each time the finalizer for a failed pod is removed and `NumRequeues` is reset, which results in a backoff of 0s.

2. Updates `updatePod` to only apply backoff when we see a particular pod failed for the first time. This is necessary to ensure that the controller does not apply backoff when it sees a pod update event for finalizer removal of a failed pod.

3. While updating the job status and removing finalizers, we returned an error to handle transient failures. Returning an error would mean that the job is re-enqueued with backoff applied. However, an error is also returned in cases when we get stale info from the informer cache, which can lead to the backoff being calculated twice. This case is now handled by checking `apierrors.IsConflict(err)` and not returning an error so that we start a fresh reconcile loop to pick up the latest version of the job from the informer cache. 

4. If `JobsReadyPods` feature is enabled, we see the 1st failed pod and backoff is 0s, the job is now enqueued after `podUpdateBatchPeriod` seconds, instead of 0s. The unit test for this check also had a few bugs:
    - `DefaultJobBackOff` is overwritten to 0 in certain unit tests, which meant that `DefaultJobBackOff` was considered to be 0, effectively not running any meaningful checks.
    - `JobsReadyPods` was not enabled for test cases that ran tests which required the feature gate to be enabled.
    - The check for expected and actual backoff had incorrect calculations.

5. Some unit tests modified `DefaultJobBackOff` but did not set it back to it's (original) default value. This commit adds `defer` statements for such tests to set it back to 10s.
    
#### Which issue(s) this PR fixes:

For https://github.com/kubernetes/kubernetes/issues/114391

#### Special notes for your reviewer:

The fix will need to be cherry-picked down to 1.23.
Update: Cherry-pick PRs are mentioned below.

- 1.23 - https://github.com/kubernetes/kubernetes/pull/115020
- 1.24 - https://github.com/kubernetes/kubernetes/pull/115021
- 1.25 - https://github.com/kubernetes/kubernetes/pull/115022
- 1.26 - https://github.com/kubernetes/kubernetes/pull/115027

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix regression in 1.25 default configurations so failed pods associated with a job with `parallelism = 1` are recreated by the job controller honoring exponential backoff delay again. However, for jobs with `parallelism > 1`, pods might be created without exponential backoff delay.
```

nikhita commented:
/assign @alculquicondor

alculquicondor commented on pkg/controller/job/job_controller.go:
```diff
@@ -879,7 +881,9 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 			// returning an error will re-enqueue Job after the backoff period
```
Did you confirm the solution in a live cluster?

note the check for `jobHasNewFailure` above. It correctly checks `failed < job.status.failed`. `job.status.failed`  is only updated when the finalizers are already removed.

I think the job is being retried fast because of `updateJob` that queues immediately. I don't think we are setting forget to true when we shouldn't.

sathyanarays replied:
I tried validating the changes in a kind cluster. I took the job definition given in the linked issue for testing. I did two runs and found similar results in both cases. The sorted pod start time are as follows.

### Run 1
2022-12-16T05:54:36Z
2022-12-16T05:54:40Z
2022-12-16T05:54:53Z
2022-12-16T05:55:16Z
2022-12-16T05:56:39Z
2022-12-16T05:59:22Z
2022-12-16T06:04:45Z
2022-12-16T06:10:49Z
2022-12-16T06:16:52Z
2022-12-16T06:22:54Z

### Run 2
2022-12-16T06:30:45Z
2022-12-16T06:30:48Z
2022-12-16T06:31:01Z
2022-12-16T06:31:24Z
2022-12-16T06:32:47Z
2022-12-16T06:35:30Z
2022-12-16T06:40:53Z
2022-12-16T06:46:57Z
2022-12-16T06:53:00Z

The time diffs in seconds for the above two run are:
Run 1: 4, 13, 23, 83, 162, 320, 363, 363, 363
Run 2: 3, 13, 23, 83, 163, 323, 364, 363

alculquicondor commented:
Thanks for confirming. It was simpler than I was thinking.
Please leave the legacy path untouched if it's not needed.
And add an integration test that verifies that the second pod created is
delayed at least 10 seconds.
We need to cherry-pick down to 1.24.

On Fri, Dec 16, 2022, 2:06 a.m. Sathyanarayanan Saravanamuthu <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pkg/controller/job/job_controller.go
> <https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1050429090>
> :
>
> > @@ -879,7 +881,9 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
>  			// returning an error will re-enqueue Job after the backoff period
>
> I tried validating the changes in a kind cluster. I took the job
> definition given in the linked issue for testing. I did two runs and found
> similar results in both cases. The sorted pod start time are as follows.
> Run 1
>
> 2022-12-16T05:54:36Z
> 2022-12-16T05:54:40Z
> 2022-12-16T05:54:53Z
> 2022-12-16T05:55:16Z
> 2022-12-16T05:56:39Z
> 2022-12-16T05:59:22Z
> 2022-12-16T06:04:45Z
> 2022-12-16T06:10:49Z
> 2022-12-16T06:16:52Z
> 2022-12-16T06:22:54Z
> Run 2
>
> 2022-12-16T06:30:45Z
> 2022-12-16T06:30:48Z
> 2022-12-16T06:31:01Z
> 2022-12-16T06:31:24Z
> 2022-12-16T06:32:47Z
> 2022-12-16T06:35:30Z
> 2022-12-16T06:40:53Z
> 2022-12-16T06:46:57Z
> 2022-12-16T06:53:00Z
>
> The time diffs in seconds for the above two run are:
> Run 1: 4, 13, 23, 83, 162, 320, 363, 363, 363
> Run 2: 3, 13, 23, 83, 163, 323, 364, 363
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1050429090>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ5E6FZAZSNJRMUZVM7OZTWNQII5ANCNFSM6AAAAAATADDRZE>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

sathyanarays commented:
> Thanks for confirming. It was simpler than I was thinking. Please leave the legacy path untouched if it's not needed. And add an integration test that verifies that the second pod created is delayed at least 10 seconds. We need to cherry-pick down to 1.24.
> […](#)
> On Fri, Dec 16, 2022, 2:06 a.m. Sathyanarayanan Saravanamuthu < ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In pkg/controller/job/job_controller.go <[#114516 (comment)](https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1050429090)> : > @@ -879,7 +881,9 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr // returning an error will re-enqueue Job after the backoff period I tried validating the changes in a kind cluster. I took the job definition given in the linked issue for testing. I did two runs and found similar results in both cases. The sorted pod start time are as follows. Run 1 2022-12-16T05:54:36Z 2022-12-16T05:54:40Z 2022-12-16T05:54:53Z 2022-12-16T05:55:16Z 2022-12-16T05:56:39Z 2022-12-16T05:59:22Z 2022-12-16T06:04:45Z 2022-12-16T06:10:49Z 2022-12-16T06:16:52Z 2022-12-16T06:22:54Z Run 2 2022-12-16T06:30:45Z 2022-12-16T06:30:48Z 2022-12-16T06:31:01Z 2022-12-16T06:31:24Z 2022-12-16T06:32:47Z 2022-12-16T06:35:30Z 2022-12-16T06:40:53Z 2022-12-16T06:46:57Z 2022-12-16T06:53:00Z The time diffs in seconds for the above two run are: Run 1: 4, 13, 23, 83, 162, 320, 363, 363, 363 Run 2: 3, 13, 23, 83, 163, 323, 364, 363 — Reply to this email directly, view it on GitHub <[#114516 (comment)](https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1050429090)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJ5E6FZAZSNJRMUZVM7OZTWNQII5ANCNFSM6AAAAAATADDRZE> . You are receiving this because you were assigned.Message ID: ***@***.***>

@alculquicondor , from what I observed, the second pod is created immediately when the first pod is created. This may happen because new failure detection and pod creation are done in the same reconcile loop. There is a 10 second gap between 2nd pod failure and 3rd pod creation.

sathyanarays commented:
Performed further testing. Details are as follows.

### Job manifest
Created a manifest that fails the pods after 1 minute
```
apiVersion: batch/v1
kind: Job
metadata:
  name: backoff-exp
spec:
  backoffLimit: 5
  completions: 25
  template:
    spec:
      containers:
        - name: die
          image: myubuntu:1
          command: ["/bin/sh","-c"]
          args: ["/bin/sleep 60 && /bin/false"]
          imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

### Start and End
The space separated container start time and end time are as follows:

```
2022-12-17T06:23:08Z 2022-12-17T06:24:08Z
2022-12-17T06:24:12Z 2022-12-17T06:25:12Z
2022-12-17T06:25:15Z 2022-12-17T06:26:15Z
2022-12-17T06:26:18Z 2022-12-17T06:27:18Z
2022-12-17T06:27:21Z 2022-12-17T06:28:21Z
2022-12-17T06:28:24Z 2022-12-17T06:29:24Z
```

After each pod failure, the next pod starts at an interval of approximately 3 seconds. I don't see exponential backoff in this case. Let me know if this makes sense!

sathyanarays commented:
> Performed further testing. Details are as follows.
> 
> ### Job manifest
> Created a manifest that fails the pods after 1 minute
> 
> ```
> apiVersion: batch/v1
> kind: Job
> metadata:
>   name: backoff-exp
> spec:
>   backoffLimit: 5
>   completions: 25
>   template:
>     spec:
>       containers:
>         - name: die
>           image: myubuntu:1
>           command: ["/bin/sh","-c"]
>           args: ["/bin/sleep 60 && /bin/false"]
>           imagePullPolicy: IfNotPresent
>       restartPolicy: Never
> ```
> 
> ### Start and End
> The space separated container start time and end time are as follows:
> 
> ```
> 2022-12-17T06:23:08Z 2022-12-17T06:24:08Z
> 2022-12-17T06:24:12Z 2022-12-17T06:25:12Z
> 2022-12-17T06:25:15Z 2022-12-17T06:26:15Z
> 2022-12-17T06:26:18Z 2022-12-17T06:27:18Z
> 2022-12-17T06:27:21Z 2022-12-17T06:28:21Z
> 2022-12-17T06:28:24Z 2022-12-17T06:29:24Z
> ```
> 
> After each pod failure, the next pod starts at an interval of approximately 3 seconds. I don't see exponential backoff in this case. Let me know if this makes sense!

Slightly changing the logic fixes this:

```
if finishedCondition != nil {
    forget = true
}
```

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -865,9 +867,9 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 	// improve the Job backoff policy when parallelism > 1 and few Jobs failed but others succeed.
 	// In this case, we should clear the backoff delay.
 	forget = job.Status.Succeeded < succeeded
+	needsStatusUpdate := suspendCondChanged || active != job.Status.Active || !equalReady(ready, job.Status.Ready) || finishedCondition != nil
```
I would suggest to only update the new path `uncounted != nil` (as also described in the Issue / PR) unless for some good reasons.

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -904,6 +908,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 		if _, err := jm.updateStatusHandler(ctx, &job); err != nil {
 			return forget, err
 		}
+
```
Revert this new line to avoid potential conflict when cherry-picking.

mimowo commented on pkg/controller/job/job_controller.go:
```diff
@@ -879,7 +881,9 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (forget bool, rEr
 			return forget, fmt.Errorf("failed pod(s) detected for job key %q", key)
 		}
-		forget = true
+		if needsStatusUpdate {
```
IIUC, the experiments done by @sathyanarays suggest there is still not expotential backoff: https://github.com/kubernetes/kubernetes/pull/114516#issuecomment-1356133958
This is probably, because in this experiments the pod is running for a while, so it enters the ready state and thus `needsStatusUpdate` is true and thus `forget=true`. This seems to actually be consistent with the legacy path (haven't tested). The suggestion by @sathyanarays is likely to fix this issue indeed: 
```golang
if finishedCondition != nil {
    forget = true
}
```
However, I'm wondering if setting `forget` when the job is failing already (due to `finishedCondition!= nil`) make any relevant change. We could probably just don't modify `forget` here and leave `true` only if there is a new succeeded pod.

alculquicondor replied:
I think we should just remove lines 884-886 completely.

nikhita replied:
@alculquicondor @mimowo maybe I'm missing something here...not sure how removing lines 884-886 would help (but agree that they won't work in the current case either)

In @sathyanarays's test case (https://github.com/kubernetes/kubernetes/pull/114516#issuecomment-1356133958), if `forget = true` is unconditionally set here, it will reset `NumRequeues` everytime and we'll always end up with a backoff of 0s.

We will also hit this case when the job status is updated to reflect the actual number of failed pods, the controller does a resync (`jobHasNewFailure` is false, `forget` is set to true) and `NumRequeues` is reset. When the _next_ failed pod shows up, it sees a backoff of 0s.

sathyanarays replied:
@alculquicondor , removing lines 884-886 would technically work. But, the job will not be marked as failed soon as the last job failure happens. The job status will be updated in the next backoff. I confirmed that with another experiment by removing lines 884-886. This may not be a good user experience.

### Details

#### Job

```
apiVersion: batch/v1
kind: Job
metadata:
  name: backoff-exp
spec:
  backoffLimit: 6
  completions: 25
  template:
    spec:
      containers:
        - name: die
          image: myubuntu:1
          command: ["/bin/sh","-c"]
          args: ["/bin/sleep 20 && /bin/false"]
          imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

#### Start and end time of pods

```
2022-12-19T13:34:46Z 2022-12-19T13:35:06Z
2022-12-19T13:35:10Z 2022-12-19T13:35:30Z
2022-12-19T13:35:53Z 2022-12-19T13:36:13Z
2022-12-19T13:36:56Z 2022-12-19T13:37:16Z
2022-12-19T13:38:39Z 2022-12-19T13:38:59Z
2022-12-19T13:41:42Z 2022-12-19T13:42:02Z
2022-12-19T13:47:26Z 2022-12-19T13:47:46Z
```
#### Job Status

```
status:
    conditions:
    - lastProbeTime: "2022-12-19T13:53:49Z"
      lastTransitionTime: "2022-12-19T13:53:49Z"
      message: Job has reached the specified backoff limit
      reason: BackoffLimitExceeded
      status: "True"
      type: Failed
    failed: 7
    ready: 0
    startTime: "2022-12-19T13:34:44Z"
    uncountedTerminatedPods: {}
```

mimowo replied:
@sathyanarays thanks for testing - it is one of the scenarios we could have integration tests (or at least fast integration tests) once the fake clock PR is merged: https://github.com/kubernetes/kubernetes/pull/110710.

alculquicondor replied:
> But, the job will not be marked as failed soon as the last job failure happens. The job status will be updated in the next backoff.

I don't think that's a regression. We do pod creations in the same cycle when we do status updates. If we want to somehow split those two actions, we could aim to do that in 1.27. But for a fix we can backport to 1.23, I think removing the lines (removing `forget=true`), is good.

But I would appreciate an integration test. Just `backoffLimit=3` should be enough.

mimowo replied:
The final status update is delayed as the last failed pod enqueues the `syncJob` with the new backoff (from within the `updatePod` function). Technically, the last job status update was always delayed like that. 

However, from the user perspective there is some regression as the job failure may be delayed by a couple of minutes (when backoffLimit=6 by default). Due to (or thanks to) the bug we are fixing the user was getting the job status much earlier. I'm leaving the decision @alculquicondor  as fixing this may be more involving indeed.

The fake clock PR is now merged so you can rebase and use. For cherry-picking we could:
1. exclude the test from being cherry-picked (during conflict resolution)
2. use a real clock in the cherry-picked test - it will make it run longer, but it should work

I'm leaning towards 2. but we can discuss it more under the cherry-pick PR.

alculquicondor replied:
Again: we were always delaying the status update, so we can do the same when tracking with finalizers.

I think for a long term solution we have to skip calling manageJob if the "timeout" is not satisfied yet. But I don't think we can just rely on `AddAfter`. We need to check within `syncJob` whether the timeout is satisfied, similar to what we do with `activeDeadlineSeconds`.

Regarding testing, I also lean towards 2.

mimowo replied:
Resetting the backoff on every job update seems to me as an artifact of the past when the updates were more rare. In particular, resetting the backoff on changing the count of pods ready seems too often.

I think this line and the comment emphasize the code intention: https://github.com/kubernetes/kubernetes/blob/1edbb8cf1a864c2dd71a1e5b7282de4bc3d8f5fa/pkg/controller/job/job_controller.go#L869-L873.

If this is the case, then I suggest to delete these lines (and adjust the unit test asserts) to only reset backoff if there is a new succeeded pod. 

This change would also make @sathyanarays PR https://github.com/kubernetes/kubernetes/pull/114768 work better as controller restart wouldn't allow us to recompute backoff properly otherwise.

On the other hand it's also ok to me to keep this PR as a minimal fix and adjust the behavior in a follow up dedicated PR. WDYT @alculquicondor ?

alculquicondor replied:
Let's special case setting `forget=true` when the job is being suspended and nothing else changed.

nikhita replied:
> If this is the case, then I suggest to delete these lines (and adjust the unit test asserts) to only reset backoff if there is a new succeeded pod.

> Let's special case setting forget=true when the job is being suspended and nothing else changed.

To double check -- does this mean that we'll special case setting `forget = true` when the job is being suspended for the legacy path as well? If we don't change the legacy path, we'll end up with different behavior for with the `JobTrackingFinalizers` feature enabled vs it being disabled.

In master, the feature is GA and locked to true so the tests don't really verify the legacy path. However, when we cherry-pick, the tests will verify for [both](https://github.com/kubernetes/kubernetes/blob/548d5d2298c4a53a775cb12163e58728f1a3bf08/pkg/controller/job/job_controller_test.go#L758) the cases (feature gate enabled/disabled). In those cases, the tests will break if both cases have different behaviors.

alculquicondor replied:
Right. Feel free to fix the legacy path for the purposes of cherry-pick.

nikhita commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -3922,15 +3944,16 @@ func (f *fakeRateLimitingQueue) AddAfter(item interface{}, duration time.Duratio
 }
 
 func TestJobBackoff(t *testing.T) {
+	DefaultJobBackOff = 10 * time.Second
```
This test was pretty broken. Made a few changes here and in job_controller.go [above](https://github.com/kubernetes/kubernetes/pull/114516/files#diff-a7577a7dede7864ff38c631319033714fdd0bed91108d976282e1099507e6ff0R509-R511).

I've included the below note in the commit message and PR body:

 If `JobsReadyPods` feature is enabled and backoff is 0s, the job is now enqueued after `podUpdateBatchPeriod` seconds, instead of 0s. The unit test for this check also had a few bugs:

- `DefaultJobBackOff` is overwritten to 0 in certain unit tests, which meant that `DefaultJobBackOff` was considered to be 0, effectively not running any meaningful checks.

- `JobsReadyPods` was not enabled for test cases that required the feature gate to b enabled.

- The check for expected and actuall backoff had incorrect calculations.

nikhita reviewed (commented):
Made a few changes and added an integration test, but I'm still not sure how to set `forget = true` correctly without hitting https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1052135114.

sathyanarays commented:
> Made a few changes and added an integration test, but I'm still not sure how to set `forget = true` correctly without hitting [#114516 (comment)](https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1052135114).

@nikhita, Can we try this?

```
if finishedCondition != nil {
    forget = true
}
```

nikhita commented:
> @nikhita, Can we try this?

> if finishedCondition != nil {
>     forget = true
> }



@sathyanarays unfortunately, that also fails unit tests:

<details>
<summary>TestControllerSyncJob failures</summary>
```
--- FAIL: TestControllerSyncJob (0.08s)
    --- FAIL: TestControllerSyncJob/suspending_a_job_with_unsatisfied_expectations,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/WQ_job:_correct_#_of_pods,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/too_few_active_pods_with_a_dynamic_job,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/correct_#_of_pods,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/limited_pods,_finalizers=true (0.01s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/WQ_job_start,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/too_many_active_pods,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/more_active_pods_than_parallelism,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/suspending_a_deleted_job,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/pending_pods,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/correct_#_of_pods,_ready_enabled,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/indexed_job_start,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/suspending_a_job_with_satisfied_expectations,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/resuming_a_suspended_job,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
    --- FAIL: TestControllerSyncJob/job_start,_finalizers=true (0.00s)
        job_controller_test.go:823: Unexpected forget value. Expected true, saw false
```
</details>

sathyanarays commented:
@nikhita, @alculquicondor, @mimowo: created an alternate fix for this issue in #114768 and it needs some refactoring. Please provide feedback if this solution is a feasible one. If yes, can this be cherry-picked to previous releases?

mimowo commented:
> created an alternate fix for this issue in #114768 and it needs some refactoring. Please provide feedback if this solution is a feasible one. If yes, can this be cherry-picked to previous releases?

It seems [PR](https://github.com/kubernetes/kubernetes/pull/114768) is the proper long-term solution and it will make it easier to recompute the backoff after controller restart, but it seems way more involving than this PR. Aside from refactoring and tests some aspects need to be fixed, for example, making sure the in-memory map does not grow uncontrolled over time. 

I think it makes sense to finish this PR first and continue working on https://github.com/kubernetes/kubernetes/pull/114768. Also, the integration test added here will be useful to have when refactoring. 

We could almost certainly cherry-pick this PR, cherry-picking the [PR](https://github.com/kubernetes/kubernetes/pull/114768) would depend on its final shape, but is less likely IMO. @alculquicondor ?

nikhita commented:
@mimowo thanks for the feedback! addressed comments to the integration test. PTAL.

mimowo commented on test/integration/job/job_test.go:
```diff
@@ -1204,6 +1205,80 @@ func TestFinalizersClearedWhenBackoffLimitExceeded(t *testing.T) {
+	defer cancel()
+
+	jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
```
Setting finalizer does not seem needed as it should be added by default now. I guess it could be an issue when cherry-picking, but then it should be better to just add 
```golang
defer featuregatetesting.SetFeatureGateDuringTest(b, feature.DefaultFeatureGate, features.JobTrackingWithFinalizers, true)()
```

alculquicondor replied:
this might be needed for the cherry-pick

mimowo commented:
One more nit remark to the test. I'm also hesitant about using a fake clock for it, what do you think @alculquicondor ? It will make it harder to cherry-pick, but should be much faster then. 

And one remaining question about setting `forget` as discussed here: https://github.com/kubernetes/kubernetes/pull/114516#issuecomment-1369773096

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -370,6 +372,19 @@ func TestControllerSyncJob(t *testing.T) {
 			expectedPodPatches: 1,
 		},
+		"no new pod; possible finalizer update of failed pod": {
+			wFinalizersExclusive: pointer.Bool(true),
```
this line is probably not needed since you fixed the legacy path too?

nikhita replied:
To make this work for the legacy path, we'll need to set `failedPods` to 1.

But setting `failedPods` to 1 would break the case with finalizers because of how failed pods are created - https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1060931512

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -209,6 +209,8 @@ func TestControllerSyncJob(t *testing.T) {
 		// If set, it means that the case is exclusive to tracking with/without finalizers.
 		wFinalizersExclusive *bool
+		// If set, it sets job.Status.Failed = expectedFailed and job.Status.Active = expectedActive
+		presetJobStatus bool
```
nit:

```suggestion
		statusCountersUpToDate bool
```

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -746,6 +761,13 @@ func TestControllerSyncJob(t *testing.T) {
 				job := newJob(tc.parallelism, tc.completions, tc.backoffLimit, tc.completionMode)
 				job.Spec.Suspend = pointer.Bool(tc.suspend)
+				if tc.presetJobStatus {
+					job.Status.Failed = tc.expectedFailed
```
nit: base it on the input, rather than the expectations

```suggestion
					job.Status.Failed = tc.failedPods
```

alculquicondor replied:
also count succeeded?

nikhita replied:
`failedPods` creates failed pods with the tracking finalizer so the job ends up thinking that the failed pod has not been accounted for in the status and ends up incrementing `job.Status.Failed`. So in this case, pre-setting the status to `failedPods` does not work.

alculquicondor replied:
ah, good point. In general it is beneficial for tests to be more verbose (following the DAMP principle). In this sense, we could actually have an explicit `initialStatus` object. But this recommendation is on the nit side.

alculquicondor commented on test/integration/job/job_test.go:
```diff
@@ -1204,6 +1205,79 @@ func TestFinalizersClearedWhenBackoffLimitExceeded(t *testing.T) {
 
+func TestJobPodsCreatedWithExponentialBackoff(t *testing.T) {
+	// overwrite the default value for faster testing
+	jobcontroller.DefaultJobBackOff = 2 * time.Second
```
defer setting the default back to it's original value.

alculquicondor commented on test/integration/job/job_test.go:
```diff
@@ -1775,6 +1865,12 @@ func getCompletionIndex(p *v1.Pod) (int, error) {
 }
 
+func failPod(ctx context.Context, t *testing.T, clientSet clientset.Interface, jobObj *batchv1.Job) {
+	if err, _ := setJobPodsPhase(ctx, clientSet, jobObj, v1.PodFailed, 1); err != nil {
```
maybe just inlining these lines is not too bad. Otherwise, worth calling t.Helper()

alculquicondor commented:
/lgtm
/approve

/hold
(feel free to remove the hold if you don't agree with my nit)

also, worth doing 2 more runs of the integration test suite before merging. Just call `/test pull-kubernetes-integration` after the test passes.

sathyanarays commented:
Performed one more round of testing. The summary and details are as follows

## Summary
The back-off works fine when parallelism is set to 1 with the following known issues: 
- No back-off between first pod failure and second pod creation
- The job is marked as finished some time (back-off) after the last failed pod

When parallelism is set to a value greater than 1, the back-off is not respected. I am not sure this is expected. 

## Details
### Manual tests

#### Job that fails immediately

```
apiVersion: batch/v1
kind: Job
metadata:
  name: backoff-exp
spec:
  backoffLimit: 25
  completions: 25
  parallelism: 1
  template:
    spec:
      containers:
        - name: die
          image: myubuntu:1
          command: ["/bin/sh","-c"]
          #args: ["/bin/sleep 1 && /bin/false"]
          args: ["/bin/false"]
          imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

There is no back-off between first failed pod and the second pod. The second and third pod has 10s; the third and fourth pod has 20s and so on.

```
2023-01-04T00:45:22Z | 2023-01-04T00:45:22Z | NA
2023-01-04T00:45:27Z | 2023-01-04T00:45:27Z | 5
2023-01-04T00:45:40Z | 2023-01-04T00:45:40Z | 13
2023-01-04T00:46:03Z | 2023-01-04T00:46:03Z | 23
2023-01-04T00:46:46Z | 2023-01-04T00:46:46Z | 43
2023-01-04T00:48:09Z | 2023-01-04T00:48:09Z | 83
2023-01-04T00:50:52Z | 2023-01-04T00:50:52Z | 163
2023-01-04T00:56:16Z | 2023-01-04T00:56:16Z | 324
2023-01-04T01:02:19Z | 2023-01-04T01:02:19Z | 363
```

#### Job that fails after 60 seconds

```
apiVersion: batch/v1
kind: Job
metadata:
  name: backoff-exp
spec:
  backoffLimit: 25
  completions: 25
  parallelism: 1
  template:
    spec:
      containers:
        - name: die
          image: myubuntu:1
          command: ["/bin/sh","-c"]
          args: ["/bin/sleep 60 && /bin/false"]
          imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

Works similar to previous case

```
2023-01-04T01:05:53Z | 2023-01-04T01:06:53Z | NA
2023-01-04T01:06:57Z | 2023-01-04T01:07:57Z | 4
2023-01-04T01:08:11Z | 2023-01-04T01:09:11Z | 14
2023-01-04T01:09:33Z | 2023-01-04T01:10:33Z | 22
2023-01-04T01:11:17Z | 2023-01-04T01:12:17Z | 44
2023-01-04T01:13:40Z | 2023-01-04T01:14:40Z | 83
2023-01-04T01:17:23Z | 2023-01-04T01:18:23Z | 163
2023-01-04T01:23:47Z | 2023-01-04T01:24:47Z | 324
2023-01-04T01:30:49Z | 2023-01-04T01:31:49Z | 362
```

#### Job with parallelism 5

```
apiVersion: batch/v1
kind: Job
metadata:
  name: backoff-exp
spec:
  backoffLimit: 25
  completions: 25
  parallelism: 5
  template:
    spec:
      containers:
        - name: die
          image: myubuntu:1
          command: ["/bin/sh","-c"]
          args: ["/bin/sleep 1 && /bin/false"]
          imagePullPolicy: IfNotPresent
      restartPolicy: Never
```

The back-off did not work. The job has backoff limit set to 25 and parallelism set to 5. The job completed within a minute.

```
  status:
    conditions:
    - lastProbeTime: "2023-01-04T02:36:36Z"
      lastTransitionTime: "2023-01-04T02:36:36Z"
      message: Job has reached the specified backoff limit
      reason: BackoffLimitExceeded
      status: "True"
      type: Failed
    failed: 30
    ready: 1
    startTime: "2023-01-04T02:35:51Z"
    uncountedTerminatedPods: {}
```

sathyanarays commented on test/integration/job/job_test.go:
```diff
@@ -1204,6 +1205,87 @@ func TestFinalizersClearedWhenBackoffLimitExceeded(t *testing.T) {
+	}
+	if creationTime[3].Sub(creationTime[2]).Seconds() < 2*jobcontroller.DefaultJobBackOff.Seconds() {
+		t.Fatalf("Fourth pod should be created at least %v seconds after the third pod", 2*jobcontroller.DefaultJobBackOff)
+	}
```
Can we also check for upper bounds by ensuring that the back off does not go beyond an acceptable threshold?

sathyanarays commented on test/integration/job/job_test.go:
```diff
@@ -1204,6 +1205,87 @@ func TestFinalizersClearedWhenBackoffLimitExceeded(t *testing.T) {
+	// so the second pod is created immediately after the first pod fails.
+	// The backoff is only applied after 2nd pod failure, so we validate exponential
+	// backoff for creationTime from the 3rd pod onwards.
+	if creationTime[2].Sub(creationTime[1]).Seconds() < jobcontroller.DefaultJobBackOff.Seconds() {
```
I tried to implement this and apply an upper bound on the back-off. While doing this, I found out that the back-offs are getting double counted sometimes (approximately 50%). I made the following changes to the new test function; I made the test fail if the second back off exceeded `4 * default backoff`. The modified code snippet is as follows:

```
	creationTime := []time.Time{}
	finishTime := []time.Time{}
	for _, pod := range jobPods {
		creationTime = append(creationTime, pod.CreationTimestamp.Time)
		if len(pod.ManagedFields) > 1 {
			finishTime = append(finishTime, pod.ManagedFields[1].Time.Time)
		}
	}

	sort.Slice(creationTime, func(i, j int) bool {
		return creationTime[i].Before(creationTime[j])
	})

	sort.Slice(finishTime, func(i, j int) bool {
		return finishTime[i].Before(finishTime[j])
	})

	// Pod failure detection and new pod creation are done in the same reconcile loop
	// so the second pod is created immediately after the first pod fails.
	// The backoff is only applied after 2nd pod failure, so we validate exponential
	// backoff for creationTime from the 3rd pod onwards.
	if creationTime[2].Sub(finishTime[1]).Seconds() < jobcontroller.DefaultJobBackOff.Seconds() {
		t.Fatalf("Third pod should be created at least %v seconds after the second pod", jobcontroller.DefaultJobBackOff)
	}

	diff := creationTime[3].Sub(finishTime[2]).Seconds()
	if diff < 2*jobcontroller.DefaultJobBackOff.Seconds() || diff >= 4*jobcontroller.DefaultJobBackOff.Seconds() {
		t.Fatalf("Fourth pod should be created at least %v seconds after the third pod", 2*jobcontroller.DefaultJobBackOff)
	}
}
```

**Note:** I have repeated the tests with different defaults: 2s, 4s and 10s. Double counting happens in all the cases.

nikhita replied:
Reproduced the backoff getting double counted on a kind cluster. Looks like it happens due to:

https://github.com/kubernetes/kubernetes/blob/575616cc72dbfdd070ead81ec29c0d4f00226487/pkg/controller/job/job_controller.go#L879-L881

Returning an error here means that the job is re-enqueued => which affects `NumRequeues` => which affects the backoff period.

Looks like we need to re-enqueue the job but not have it affect the backoff period...

sathyanarays replied:
Some more details into this.

In some reconcile loops, the job status is updated twice. An example is as follows:

#### First one
```
{"startTime":"2023-01-05T08:47:34Z","active":1,"failed":3,"uncountedTerminatedPods":{"failed":["88c93042-2d76-4214-bad0-23842e3caa19"]},"ready":0}
```

#### Second one
```
{"startTime":"2023-01-05T08:47:34Z","active":1,"failed":4,"uncountedTerminatedPods":{},"ready":0}
```

In the next reconcile loop, the status of the job queried from the informer is as follows and it is equal to the first update.

#### Status as queried from informer
```
{"startTime":"2023-01-05T08:47:34Z","active":1,"failed":3,"uncountedTerminatedPods":{"failed":["88c93042-2d76-4214-bad0-23842e3caa19"]},"ready":0}
```

In the same loop, if we query from API server (by passing informer cache), we get the following

```
{"startTime":"2023-01-05T08:47:34Z","active":1,"failed":4,"uncountedTerminatedPods":{},"ready":0}
```

In short, this happens due to staleness of the informer cache and it is expected to have stale caches under load.

sathyanarays replied:
The following solution seems to be working. I am not encountering the double back-off after adding these lines. Please let me know your thoughts! 

```
		err = jm.trackJobStatusAndRemoveFinalizers(ctx, &job, pods, prevSucceededIndexes, *uncounted, expectedRmFinalizers, finishedCondition, needsStatusUpdate)
		if err != nil { // assuming all errors here will be concurrent modification errors; we may do this error specific
			jm.enqueueController(job, false)
			return false, nil
		}

```

alculquicondor replied:
But we want backoff if there are API failures (to avoid clogging the apiserver). This backoff should be independent of the backoff for pod failures, but again, that's something that we should solve in https://github.com/kubernetes/kubernetes/pull/114768

sathyanarays replied:
@alculquicondor , This particular error that we face (described in comment [1062243513](https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1062243513)) is not transient that goes away with retries in the same reconcile loop. We should come out of the reconcile loop to start a fresh loop that picks up the latest version of the job from the informer cache.

For other transient failures, I agree with your point!

alculquicondor replied:
Ah right, the stale info in the informer. You can specifically detect this using `errors.IsConflict`. I'm ok special casing this.

sathyanarays commented:
In addition to the issues described in comment [1370452470](https://github.com/kubernetes/kubernetes/pull/114516#issuecomment-1370452470), we also see back-off being double counted in some instances. It can be confirmed by modifying the integration test as in comment [1061270734](https://github.com/kubernetes/kubernetes/pull/114516#discussion_r1061270734).

nikhita commented:
Going to cancel the lgtm to fix all the issues.

/lgtm cancel

alculquicondor commented:
/approve cancel

alculquicondor commented:
I think the issue might be that each pod creation (before it fails) triggers an immediate sync.

I'm 90% confident that this was always a problem, so this can't be treated as a regression. If you see value on fixing the backoffs for parallelism=1, I'm happy to proceed with this PR after the integration test is fixed as suggested by @sathyanarays (and cherry-pick).

The long term solution should be https://github.com/kubernetes/kubernetes/pull/114768, but I think we should keep it in 1.27+

/kind bug
/remove-kind regression

nikhita commented:
@alculquicondor @sathyanarays @mimowo addressed comments.

But I've noticed the integration test acting flaky locally so the PR is not ready to be merged yet. Will debug and fix the test but pushed changes to address comments for early feedback.

As a heads up, I'm on PTO early next week but will try to wrap up by EOW.

sathyanarays commented:
> @alculquicondor @sathyanarays @mimowo addressed comments.
> 
> But I've noticed the integration test acting flaky locally so the PR is not ready to be merged yet. Will debug and fix the test but pushed changes to address comments for early feedback.
> 
> As a heads up, I'm on PTO early next week but will try to wrap up by EOW.

I had a look at the issue of pods getting scheduled immediately on 3rd or 4th retries. This looks like a concurrency issue. There are many possibilities for this to happen. I observed one such caused by the following sequences of events.

#### Sequence 1 (Reconcile loop)
Initial State:
- Job status: active=1
- Pod status: active=0; failed=1

Events:
1. Sync Job called (Reconcile loop N)
1. Compare Pod Status and Job status
2. Pod Status and Job status differ, so create a pod (Pod created here concurrent sequence of events as described in sequence 2)
3. Update Job
4. Complete Reconcile loop

#### Sequence 2 
1. Pod gets created (asynchronously triggers (Triggers sequence 3)
2. After some time(almost immediately), the pod's status set to fail by the integration test (Triggers sequence 4)

#### Sequence 3
1. The pod informer cache gets notified of the pod creation & cache gets updated (Sequence 2, event 1)
2. The pod creation handler gets called
3. Wait for 1 second
4. Add the message to workqueue
5. Sync Job called (Reconcile loop N+k)
6. Pod status and Job status are compared (E1)

#### Sequence 4
1. The pod informer cache gets notified about the pod update & cache gets updated(sequence 2, event 2) (E2)

Given the above concurrent sequences of events, we are interested in two events: E1 => (Sequence 3, Event 6) and E2 => (Sequence 4, Event 1).

When **E2 happens before E1**, the integration tests pass and we are good.
When **E1 happens before E2**, the reconcile loop N+k finds difference between pod status and job status; it triggers creation of new pod(s)

This is one possible sequence of events that makes the integration test fail. There can be other sequences causing the same effect.

sathyanarays commented:
> > @alculquicondor @sathyanarays @mimowo addressed comments.
> > But I've noticed the integration test acting flaky locally so the PR is not ready to be merged yet. Will debug and fix the test but pushed changes to address comments for early feedback.
> > As a heads up, I'm on PTO early next week but will try to wrap up by EOW.
> 
> I had a look at the issue of pods getting scheduled immediately on 3rd or 4th retries. This looks like a concurrency issue. There are many possibilities for this to happen. I observed one such caused by the following sequences of events.
> 
> #### Sequence 1 (Reconcile loop)
> Initial State:
> 
> * Job status: active=1
> * Pod status: active=0; failed=1
> 
> Events:
> 
> 1. Sync Job called (Reconcile loop N)
> 2. Compare Pod Status and Job status
> 3. Pod Status and Job status differ, so create a pod (Pod created here concurrent sequence of events as described in sequence 2)
> 4. Update Job
> 5. Complete Reconcile loop
> 
> #### Sequence 2
> 1. Pod gets created (asynchronously triggers (Triggers sequence 3)
> 2. After some time(almost immediately), the pod's status set to fail by the integration test (Triggers sequence 4)
> 
> #### Sequence 3
> 1. The pod informer cache gets notified of the pod creation & cache gets updated (Sequence 2, event 1)
> 2. The pod creation handler gets called
> 3. Wait for 1 second
> 4. Add the message to workqueue
> 5. Sync Job called (Reconcile loop N+k)
> 6. Pod status and Job status are compared (E1)
> 
> #### Sequence 4
> 1. The pod informer cache gets notified about the pod update & cache gets updated(sequence 2, event 2) (E2)
> 
> Given the above concurrent sequences of events, we are interested in two events: E1 => (Sequence 3, Event 6) and E2 => (Sequence 4, Event 1).
> 
> When **E2 happens before E1**, the integration tests pass and we are good. When **E1 happens before E2**, the reconcile loop N+k finds difference between pod status and job status; it triggers creation of new pod(s)
> 
> This is one possible sequence of events that makes the integration test fail. There can be other sequences causing the same effect.

I slightly modified the integration test to have `time.Sleep(time.Second)` before each `setJobPodsPhase` call and ran the test 1000 times. The test passed all 1000 times. This adds more evidence (not proof) to the above hypotheses.

alculquicondor commented:
The sleep you suggest reduces the chances that `syncJob` gets triggered for a pod creation (`immediate=true`), but the pod fails before `syncJob` actually counts the pods.

I'm not too keen on having a sleep in the integration test. I think we could just merge with the unit tests and leave the integration test for the final fix (where we separate pod creation backoff from the workqueue delays).

sathyanarays commented:
> The sleep you suggest reduces the chances that `syncJob` gets triggered for a pod creation (`immediate=true`) but the pod fails before `syncJob` actually counts the pods.

@alculquicondor , I am not suggesting adding sleep to the integration test. I just used it to illustrate that if the pod runs at least one second before it fails, then everything works fine. We will see the issue of the pod getting scheduled immediately (3rd, 4th,...) if the pod fails within a second. The fix for this should go in the `job_controller.go` and not in the integration tests.
 
> I'm not too keen on having a sleep in the integration test. I think we could just merge with the unit tests and leave the integration test for the final fix (where we separate pod creation backoff from the workqueue delays).

+1

nikhita committed:
pkg/controller/job: re-honor exponential backoff

This commit makes the job controller re-honor exponential backoff for
failed pods. Before this commit, the controller created pods without any
backoff. This is a regression because the controller used to
create pods with an exponential backoff delay before (10s, 20s, 40s ...).

The issue occurs only when the JobTrackingWithFinalizers feature is
enabled (which is enabled by default right now). With this feature, we
get an extra pod update event when the finalizer of a failed pod is
removed.

Note that the pod failure detection and new pod creation happen in the
same reconcile loop so the 2nd pod is created immediately after the 1st
pod fails. The backoff is only applied on 2nd pod failure, which means
that the 3rd pod created 10s after the 2nd pod, 4th pod is created 20s
after the 3rd pod and so on.

This commit fixes a few bugs:

1. Right now, each time `uncounted != nil` and the job does not see a
_new_ failure, `forget` is set to true and the job is removed from the
queue. Which means that this condition is also triggered each time the
finalizer for a failed pod is removed and `NumRequeues` is reset, which
results in a backoff of 0s.

2. Updates `updatePod` to only apply backoff when we see a particular
pod failed for the first time. This is necessary to ensure that the
controller does not apply backoff when it sees a pod update event
for finalizer removal of a failed pod.

3. If `JobsReadyPods` feature is enabled and backoff is 0s, the job is
now enqueued after `podUpdateBatchPeriod` seconds, instead of 0s. The
unit test for this check also had a few bugs:
    - `DefaultJobBackOff` is overwritten to 0 in certain unit tests,
    which meant that `DefaultJobBackOff` was considered to be 0,
    effectively not running any meaningful checks.
    - `JobsReadyPods` was not enabled for test cases that ran tests
    which required the feature gate to be enabled.
    - The check for expected and actual backoff had incorrect
    calculations.

nikhita commented:
>  I think we could just merge with the unit tests and leave the integration test for the final fix (where we separate pod creation backoff from the workqueue delays).

Removed the integration test.

alculquicondor commented on pkg/controller/job/job_controller_test.go:
```diff
@@ -4043,47 +4084,47 @@ func TestJobBackoffForOnFailure(t *testing.T) {
 	}{
 		"backoffLimit 0 should have 1 pod active": {
```
None of the tests have `forget=true` now. Can you add some?

alculquicondor commented:
Should we add a sentence in the release note that a higher level of parallelism might lead to pods being recreated faster?

/lgtm
/approve
/hold
for thoughts on the release note.

nikhita commented:
> Should we add a sentence in the release note that a higher level of parallelism might lead to pods being recreated faster?

Updated the release note and also created cherry-pick PRs:
- 1.23 - https://github.com/kubernetes/kubernetes/pull/115020
- 1.24 - https://github.com/kubernetes/kubernetes/pull/115021
- 1.25 - https://github.com/kubernetes/kubernetes/pull/115022

alculquicondor commented:
> pods are created without exponential backoff delay.

"might be"? We can't guarantee either behavior for now.

nikhita commented:
> "might be"? We can't guarantee either behavior for now.

Updated the release note in this PR + all cherry-pick PRs.

> You are missing a cherry pick for 1.26

oops, created https://github.com/kubernetes/kubernetes/pull/115027

nikhita commented:
@alculquicondor the lgtm was removed due to the rebase. Can you re-apply? Thanks!

alculquicondor commented:
/lgtm
/hold cancel
Thanks!
